### PR TITLE
Add Badger DAO subgraph

### PR DESCRIPTION
**Pull request by AP-Atul**
- Using [schema-yield.graphql](https://github.com/messari/subgraphs/blob/master/schema-yield.graphql)

- Note: The token price in USD is only readable after block 13M, before that block number `tokenPrice` is always 0.

### CODE REVIEW

File: subgraphs/badger-dao/src/handlers/deposit.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import { getOrCreateDeposit } from "../entities/Transaction";
+import { getPriceOfStakedTokens, getUsdPriceOfToken } from "./price";
+
+export function deposit(call: ethereum.Call, vault: Vault, amount: BigInt): void {
+  let deposit = getOrCreateDeposit(call.transaction.hash, call.transaction.index);
+
+  let totalSupply = vault.outputTokenSupply;
+  let balance = vault.inputTokenBalances[0];
+
+  let shares = totalSupply.isZero() ? amount : amount.times(totalSupply).div(balance);
```
Comment (harsh9200) [inline subgraphs/badger-dao/src/handlers/deposit.ts:13]: I think there is some logical issue in how you are calculating the shares minted. There are some logs in your subgraph where the deposit is negative which is wrong.

image

File: subgraphs/badger-dao/src/handlers/deposit.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  vault.totalVolumeUSD = vault.totalVolumeUSD.plus(
+    inputTokenPrice.times(amount.toBigDecimal()).div(inputTokenDecimals.toBigDecimal()),
+  );
+
+  vault.outputTokenPriceUSD = getPriceOfStakedTokens(
+    Address.fromString(vault.id),
+    inputTokenAddress,
+    inputTokenDecimals,
+  ).toBigDecimal();
```
Comment (harsh9200) [inline subgraphs/badger-dao/src/handlers/deposit.ts:32]: This value will be zero if the result is less than one cause you are doing all the arithmetic in BigInt. Change this to BigDecimal.

**Comment by harsh9200** [comment]
Hello @AP-Atul, I have gone through your subgraph and the `withdraws` and `deposits` entities are not properly linked to the `vault` entity. The value of `totalValueLockedUSD` is negative in some entities which I think is due to the fact that some of your deposits and withdrawal amounts are negative. 
 
 This is one of the problematic responses that I got when I queried your subgraph.
 
 ```
 {
        "id": "0x2b5455aac8d64c14786c3a29858e43b5945819c0",
        "inputTokens": [
          {
            "id": "0x62b9c7356a2dc64a1969e19c23e4f579f9810aa7",
            "name": "Convex CRV",
            "symbol": "cvxCRV",
            "decimals": 18
          }
        ],
        "outputToken": {
          "id": "0x2b5455aac8d64c14786c3a29858e43b5945819c0",
          "name": "Badger Sett Convex CRV",
          "symbol": "bcvxCRV",
          "decimals": 18
      },
      "totalValueLockedUSD": "32242646.716810195533700310401018",
      "totalVolumeUSD": "-44892730.549127507658901940165573",
      "inputTokenBalances": ["11601476944924442195057071"],
      "outputTokenSupply": "-403774488479015730829158",
      "outputTokenPriceUSD": "0",
      "createdTimestamp": "1624373395",
      "createdBlockNumber": "12684635",
      "name": "Badger Sett Convex CRV",
      "symbol": "bcvxCRV",
      "deposits": [],
      "withdraws": []
 }
```

File: subgraphs/badger-dao/src/entities/Metrics.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+export function getOrCreateUserSnapshot(day: i32): UsageMetricsDailySnapshot {
+  let snapshot = UsageMetricsDailySnapshot.load(day.toString());
+
+  if (snapshot) {
+    return snapshot;
+  }
+
+  snapshot = new UsageMetricsDailySnapshot(day.toString());
+
+  snapshot.protocol = "";
```
Comment (this-username-is-taken) [inline subgraphs/badger-dao/src/entities/Metrics.ts:15]: why empty string?

File: subgraphs/badger-dao/subgraph.template.yaml [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  - kind: ethereum/contract
+    name: {{ name }}
+    network: {{ chain }}
+    source:
+      address: "{{ address }}"
+      abi: BadgerController
+      startBlock: {{ startBlock }}
+    mapping:
+      kind: ethereum/events
+      apiVersion: 0.0.5
```
Comment (this-username-is-taken) [inline subgraphs/badger-dao/subgraph.template.yaml:15]: update to 0.0.6

File: subgraphs/badger-dao/src/entities/Protocol.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  const id = PROTOCOL_ID.toHex();
+  let protocol = YieldAggregator.load(id);
+
+  if (protocol) {
+    return protocol;
+  }
+
+  protocol = new YieldAggregator(id);
+
+  // TODO: values to verify
```
Comment (this-username-is-taken) [inline subgraphs/badger-dao/src/entities/Protocol.ts:14]: is this still needed?

File: subgraphs/badger-dao/src/entities/Protocol.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+export function getOrCreateProtocol(): YieldAggregator {
+  const id = PROTOCOL_ID.toHex();
+  let protocol = YieldAggregator.load(id);
+
+  if (protocol) {
+    return protocol;
+  }
+
+  protocol = new YieldAggregator(id);
```
Comment (this-username-is-taken) [inline subgraphs/badger-dao/src/entities/Protocol.ts:12]: add `schemaVersion` and `subgraphVersion`

File: subgraphs/badger-dao/src/handlers/deposit.ts (hunk 1 above)
Comment (AP-Atul) [reply subgraphs/badger-dao/src/handlers/deposit.ts:13]: > Hello @AP-Atul, I have gone through your subgraph and the `withdraws` and `deposits` entities are not properly linked to the `vault` entity. The value of `totalValueLockedUSD` is negative in some entities which I think is due to the fact that some of your deposits and withdrawal amounts are negative.
> 
> This is one of the problematic responses that I got when I queried your subgraph.
> 
> ```
> {
>    "id": "0x2b5455aac8d64c14786c3a29858e43b5945819c0",
>    "inputTokens": [
>      {
>        "id": "0x62b9c7356a2dc64a1969e19c23e4f579f9810aa7",
>        "name": "Convex CRV",
>        "symbol": "cvxCRV",
>        "decimals": 18
>      }
>    ],
>    "outputToken": {
>      "id": "0x2b5455aac8d64c14786c3a29858e43b5945819c0",
>      "name": "Badger Sett Convex CRV",
>      "symbol": "bcvxCRV",
>      "decimals": 18
>    },
>    "totalValueLockedUSD": "32242646.716810195533700310401018",
>    "totalVolumeUSD": "-44892730.549127507658901940165573",
>    "inputTokenBalances": ["11601476944924442195057071"],
>    "outputTokenSupply": "-403774488479015730829158",
>    "outputTokenPriceUSD": "0",
>    "createdTimestamp": "1624373395",
>    "createdBlockNumber": "12684635",
>    "name": "Badger Sett Convex CRV",
>    "symbol": "bcvxCRV",
>    "deposits": [],
>    "withdraws": []
> }
> ```

I have linked the withdraw and deposit entities.

Comment (AP-Atul) [reply subgraphs/badger-dao/src/handlers/deposit.ts:13]: And for negative amount, trying to mitigate with checks

**Comment by AP-Atul** [comment]
Using `Transfer` event on the Sett/Vault to determine deposit or withdraw action
[reference](https://github.com/Badger-Finance/badger-subgraph/blob/b281b1e83886dbde84aaa001f9debac0b117adf8/src/vaults/src/handlers/sett-handler.ts#L95)

**Pull request closed without merging**